prov/util/src/util_attr.c

```diff
@@ -426,43 +426,76 @@ static int fi_resource_mgmt_level(enum fi_resource_mgmt rm_model)
 /* If a provider supports basic registration mode it should set the FI_MR_BASIC
  * mode bit in prov_mode. Support for FI_MR_SCALABLE is indicated by not setting
  * any of OFI_MR_BASIC_MAP bits. */
-int ofi_check_mr_mode(uint32_t api_version, uint32_t prov_mode,
+int ofi_check_mr_mode(const struct fi_provider *prov, uint32_t api_version,
+		      uint64_t user_info_caps, uint32_t prov_mode,
 		      uint32_t user_mode)
 {
+	uint64_t prov_mode_log;
+	int ret;
+
 	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
+		if (!(user_info_caps & (FI_RMA | FI_ATOMIC)) &&
+		    !(prov_mode & FI_MR_LOCAL))
+			return 0;
+
 		prov_mode &= ~FI_MR_LOCAL; /* ignore local bit */
 
 		switch (user_mode) {
 		case FI_MR_UNSPEC:
-			return OFI_CHECK_MR_SCALABLE(prov_mode) ||
+			ret = OFI_CHECK_MR_SCALABLE(prov_mode) ||
 				OFI_CHECK_MR_BASIC(prov_mode) ?
 				0 : -FI_ENODATA;
+			break;
 		case FI_MR_BASIC:
-			return OFI_CHECK_MR_BASIC(prov_mode) ? 0 : -FI_ENODATA;
+			ret = OFI_CHECK_MR_BASIC(prov_mode) ? 0 : -FI_ENODATA;
+			break;
 		case FI_MR_SCALABLE:
-			return OFI_CHECK_MR_SCALABLE(prov_mode) ? 0 : -FI_ENODATA;
+			ret = OFI_CHECK_MR_SCALABLE(prov_mode) ? 0 : -FI_ENODATA;
+			break;
 		default:
-			return -FI_ENODATA;
+			ret = -FI_ENODATA;
+			break;
 		}
 	} else {
+		if (!(user_info_caps & (FI_RMA | FI_ATOMIC))) {
+			if (prov_mode & FI_MR_LOCAL)
+				prov_mode &= ~(FI_MR_PROV_KEY | FI_MR_VIRT_ADDR);
+			else
+				prov_mode &= ~OFI_MR_BASIC_MAP;
+		}
+
 		if (user_mode & FI_MR_BASIC) {
 			if (!OFI_CHECK_MR_BASIC(prov_mode))
-				return -FI_ENODATA;
-			if ((user_mode & prov_mode & ~OFI_MR_BASIC_MAP) ==
-			    (prov_mode & ~OFI_MR_BASIC_MAP))
-				return 0;
-			return -FI_ENODATA;
+				ret = -FI_ENODATA;
+			else if ((user_mode & prov_mode & ~OFI_MR_BASIC_MAP) ==
+				 (prov_mode & ~OFI_MR_BASIC_MAP))
+				ret = 0;
+			else
+				ret = -FI_ENODATA;
 		} else {
-			return (((user_mode | FI_MR_BASIC) & prov_mode) == prov_mode) ?
-				0 : -FI_ENODATA;
+			ret = (((user_mode | FI_MR_BASIC) & prov_mode) ==
+			       prov_mode) ? 0 : -FI_ENODATA;
 		}
 	}
+
+	if (ret) {
+		FI_INFO(prov, FI_LOG_CORE, "Invalid memory registration mode\n");
+		if (FI_VERSION_GE(api_version, FI_VERSION(1, 5)))
+			prov_mode_log = prov_mode & ~(FI_MR_BASIC | FI_MR_SCALABLE);
```

Comment:
Wouldn't it be more useful to log the original registration mode?

Reply:
If we had zeroed OFI_MR_BASIC_MAP mode bits in prov_mode because the app hadn't requested RMA/Atomics there is no need to log it as required when the failure is due to some other mode bit. Also FI_MR_BASIC and FI_MR_SCALABLE bits (if present) are zeroed since the app version is >= 1.5